api/src/main/java/org/openmrs/module/locationbasedaccess/PatientSearchAdvise.java

```diff
@@ -0,0 +1,96 @@
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ * <p>
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+
+package org.openmrs.module.locationbasedaccess;
+
+
+import org.aopalliance.aop.Advice;
+import org.aopalliance.intercept.MethodInvocation;
+import org.aopalliance.intercept.MethodInterceptor;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.openmrs.Location;
+import org.openmrs.Patient;
+import org.openmrs.Person;
+import org.openmrs.PersonAttribute;
+import org.openmrs.api.context.Context;
+import org.springframework.aop.Advisor;
+import org.springframework.aop.support.StaticMethodMatcherPointcutAdvisor;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * It will be used to implement the AOP Methods for PatientServiceImpl class
+ */
+public class PatientSearchAdvise extends StaticMethodMatcherPointcutAdvisor implements Advisor {
+
+    protected static final Log log = LogFactory.getLog(PatientSearchAdvise.class);
+
+    @Override
+    public boolean matches(Method method, Class targetClass) {
+        if (method.getName().equals("getPatients")) {
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public Advice getAdvice() {
+        return new PatientSearchControlAdvice();
+    }
+
+    /**
+     * This inner class will be used to intercept the caught method using AOP
+     */
+    private class PatientSearchControlAdvice implements MethodInterceptor {
+
+        public Object invoke(MethodInvocation invocation) throws Throwable {
+            Integer sessionLocationId = Context.getUserContext().getLocationId();
+            log.error(invocation.getMethod().getName());
+            if(sessionLocationId != null) {
+                Location sessionLocation = Context.getLocationService().getLocation(sessionLocationId);
+                List<Patient> patientList = (List<Patient>)invocation.proceed();
+                for (Iterator<Patient> iterator = patientList.iterator(); iterator.hasNext(); ) {
+                    Patient patient  = iterator.next();
+                    patient = isPatientInLocation(patient, sessionLocation);
+                    if(patient == null) {
+                        iterator.remove();
+                    }
+                }
+                return patientList;
+            }
+            else {
+                // If session location doesn't available, return empty patient list
+                log.debug("Search Patient : Null Session Location in the UserContext");
+                return new ArrayList<Patient>();
+            }
+        }
+
+        private  Patient isPatientInLocation(Patient patient, Location sessionLocation) {
```

Comment:
The parameters and contents of this method should not know anything about this location being for session.

Reply:
Sorry I didn't get you properly. Can you please explain in more detail?

Reply:
In other words, what is the motivation for including "session" in this method?

Reply:
I didn't get you properly. Anyway, there are two ways to get it, 
* Actually, we don't need to pass this location through the method parameters. We can fetch it inside this method also. I thought, Getting sessionLocationUUId, and validating that Uuid, and then getting location using that UUID may require some amount of time, So if we do it for every patient condition(inside this isPatientLocation method), it may take some amount of time. So better to fetch it once per search, and pass it over every patient. 

* Did you mean about the naming convention of this parameter :smile: ? We can change it to `location` as well.

Reply:
Names reflect what a method is about. So what has session got to do with that method?